File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/r2dbc/R2dbcRepositoryGenerator.java
package ru.tinkoff.kora.database.annotation.processor.r2dbc;

import com.palantir.javapoet.*;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.*;
import ru.tinkoff.kora.common.Tag;
import ru.tinkoff.kora.database.annotation.processor.DbUtils;
import ru.tinkoff.kora.database.annotation.processor.QueryWithParameters;
import ru.tinkoff.kora.database.annotation.processor.RepositoryGenerator;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameter;
import ru.tinkoff.kora.database.annotation.processor.model.QueryParameterParser;

import javax.annotation.processing.Filer;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public final class R2dbcRepositoryGenerator implements RepositoryGenerator {
    private final TypeMirror repositoryInterface;
    private final Types types;
    private final Elements elements;
    private final Filer filer;

    public R2dbcRepositoryGenerator(ProcessingEnvironment processingEnv) {
        var repository = processingEnv.getElementUtils().getTypeElement(R2dbcTypes.R2DBC_REPOSITORY.canonicalName());
        if (repository == null) {
            this.repositoryInterface = null;
        } else {
            this.repositoryInterface = repository.asType();
        }
        this.types = processingEnv.getTypeUtils();
        this.elements = processingEnv.getElementUtils();
        this.filer = processingEnv.getFiler();
    }

    @Nullable
    @Override
    public TypeMirror repositoryInterface() {
        return this.repositoryInterface;
    }


    @Override
    public TypeSpec generate(TypeElement repositoryElement, TypeSpec.Builder type, MethodSpec.Builder constructor) {
        var repositoryType = (DeclaredType) repositoryElement.asType();
        var queryMethods = DbUtils.findQueryMethods(this.types, this.elements, repositoryElement);
        this.enrichWithExecutor(repositoryElement, type, constructor);
        var resultMappers = new FieldFactory(this.types, elements, type, constructor, "_result_mapper_");
        var parameterMappers = new FieldFactory(this.types, elements, type, constructor, "_parameter_mapper_");

        int methodCounter = 1;
        for (var method : queryMethods) {
            var methodType = (ExecutableType) this.types.asMemberOf(repositoryType, method);
            var parameters = QueryParameterParser.parse(this.types, R2dbcTypes.CONNECTION, R2dbcTypes.PARAMETER_COLUMN_MAPPER, method, methodType);
            var queryAnnotation = AnnotationUtils.findAnnotation(method, DbUtils.QUERY_ANNOTATION);
            var queryString = AnnotationUtils.<String>parseAnnotationValueWithoutDefault(queryAnnotation, "value");
            var query = QueryWithParameters.parse(filer, types, queryString, parameters, repositoryType, method);
            var resultMapper = this.parseResultMapper(method, parameters, methodType)
                .map(rm -> DbUtils.addMapper(resultMappers, rm))
                .orElse(null);
            DbUtils.addMappers(parameterMappers, DbUtils.parseParameterMappers(
                parameters,
                query,
                tn -> R2dbcNativeTypes.findAndBox(tn) != null,
                R2dbcTypes.PARAMETER_COLUMN_MAPPER
            ));
            var methodSpec = this.generate(type, methodCounter, method, methodType, query, parameters, resultMapper, parameterMappers);
            type.addMethod(methodSpec);
            methodCounter++;
        }
        return type.addMethod(constructor.build()).build();
    }

    private MethodSpec generate(TypeSpec.Builder type, int methodNumber, ExecutableElement method, ExecutableType methodType, QueryWithParameters query, List<QueryParameter> parameters, @Nullable String resultMapperName, FieldFactory parameterMappers) {
        final boolean generatedKeys = AnnotationUtils.isAnnotationPresent(method, DbUtils.ID_ANNOTATION);
        var sql = query.rawQuery();
        for (var parameter : query.parameters().stream().sorted(Comparator.<QueryWithParameters.QueryParameter>comparingInt(s -> s.sqlParameterName().length()).reversed()).toList()) {
            for (var sqlIndex : parameter.sqlIndexes()) {
                sql = sql.replace(":" + parameter.sqlParameterName(), "$" + (sqlIndex + 1));
            }
        }
        var connectionParameter = parameters.stream().filter(QueryParameter.ConnectionParameter.class::isInstance).findFirst().orElse(null);

        var b = DbUtils.queryMethodBuilder(method, methodType);

        var queryContextFieldName = "QUERY_CONTEXT_" + methodNumber;
        type.addField(
            FieldSpec.builder(DbUtils.QUERY_CONTEXT, queryContextFieldName, Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
                .initializer("""
                    new $T(
                          $S,
                          $S,
                          $S
                        )""", DbUtils.QUERY_CONTEXT, query.rawQuery(), sql, DbUtils.operationName(method))
                .build());
        b.addStatement("var _query = $L", queryContextFieldName);

        var batchParam = parameters.stream().filter(QueryParameter.BatchParameter.class::isInstance).findFirst().orElse(null);
        var isFlux = CommonUtils.isFlux(methodType.getReturnType());
        var isMono = CommonUtils.isMono(methodType.getReturnType());

        var returnType = isMono || isFlux
            ? ((DeclaredType) method.getReturnType()).getTypeArguments().get(0)
            : method.getReturnType();

        b.addCode("var _result = ");
        b.addCode("$T.deferContextual(_reactorCtx -> {$>\n", isFlux ? CommonClassNames.flux : CommonClassNames.mono);
        b.addStatement("var _ctxCurrent = $T.current(_reactorCtx)", CommonClassNames.contextReactor);
        b.addStatement("var _ctxFork = _ctxCurrent.fork()");
        b.addStatement("_ctxFork.inject()");
        b.addCode("var _telemetry = this._connectionFactory.telemetry().createContext(_ctxFork, _query);\n");
        var connectionName = "_con";
        if (connectionParameter == null) {
            b.addCode("return this._connectionFactory.withConnection$L(_con -> {$>\n", isFlux ? "Flux" : "");
        } else {
            connectionName = connectionParameter.name();
        }
        b.addCode("var _stmt = $N.createStatement(_query.sql());\n", connectionName);

        R2dbcStatementSetterGenerator.generate(b, method, query, parameters, batchParam, parameterMappers);

        if (generatedKeys) {
            b.addCode("var _flux = $T.<$T>from(_stmt.returnGeneratedValues().execute());\n", CommonClassNames.flux, R2dbcTypes.RESULT);
        } else {
            b.addCode("var _flux = $T.<$T>from(_stmt.execute());\n", CommonClassNames.flux, R2dbcTypes.RESULT);
        }

        var mappings = CommonUtils.parseMapping(method);
        var resultFluxMapper = mappings.getMapping(R2dbcTypes.RESULT_FLUX_MAPPER);
        var rowMapper = mappings.getMapping(R2dbcTypes.ROW_MAPPER);
        if (returnType.toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
            b.addCode("return _flux.flatMap($T::getRowsUpdated).reduce(0L, Long::sum).map($T::new)", R2dbcTypes.RESULT, DbUtils.UPDATE_COUNT);
        } else if (resultFluxMapper != null) {
            Objects.requireNonNull(resultMapperName, () -> "Illegal State occurred when expected to get result mapper, but got null in " + method.getEnclosingElement().getSimpleName() + "#" + method.getSimpleName());
            b.addCode("return $L.apply(_flux)\n", resultMapperName);
        } else if (rowMapper != null || !CommonUtils.isVoid(returnType)) {
            Objects.requireNonNull(resultMapperName, () -> "Illegal State occurred when expected to get result mapper, but got null in " + method.getEnclosingElement().getSimpleName() + "#" + method.getSimpleName());
            b.addCode("return $L.apply(_flux)\n", resultMapperName);
        } else {
            b.addCode("return _flux.flatMap($T::getRowsUpdated).then()", R2dbcTypes.RESULT);
        }
        b.addCode("""
            .doOnEach(_s -> {
              if (_s.isOnComplete()) {
                _telemetry.close(null);
                _ctxCurrent.inject();
              } else if (_s.isOnError()) {
                _telemetry.close(_s.getThrowable());
                _ctxCurrent.inject();
              }
            });""");
        if (connectionParameter == null) {
            b.addCode("\n$<\n});\n");
        }
        b.addCode("\n$<\n});\n");
        if (isMono || isFlux) {
            b.addCode("return _result;\n");
        } else if (returnType.getKind() == TypeKind.VOID) {
            b.addCode("_result.then().block();");
        } else {
            b.addCode("return _result.block();");
        }
        return b.build();
    }

    private Optional<DbUtils.Mapper> parseResultMapper(ExecutableElement method, List<QueryParameter> parameters, ExecutableType methodType) {
        var returnType = methodType.getReturnType();
        final boolean isFlux = CommonUtils.isFlux(returnType);
        final boolean isMono = CommonUtils.isMono(returnType);
        final boolean generatedKeys = AnnotationUtils.isAnnotationPresent(method, DbUtils.ID_ANNOTATION);
        for (var parameter : parameters) {
            if (parameter instanceof QueryParameter.BatchParameter) {
                final TypeMirror realReturnType = (isMono || isFlux)
                    ? Visitors.visitDeclaredType(returnType, dt -> dt.getTypeArguments().get(0))
                    : method.getReturnType();

                if (CommonUtils.isVoid(realReturnType)) {
                    return Optional.empty();
                } else if (DbUtils.UPDATE_COUNT.canonicalName().equals(realReturnType.toString())) {
                    return Optional.empty();
                } else if (!generatedKeys) {
                    throw new ProcessingErrorException("@Batch method can't return arbitrary values, it can only return: void/UpdateCount or database-generated @Id", method);
                }
            }
        }

        var mappings = CommonUtils.parseMapping(method);
        var resultFluxMapper = mappings.getMapping(R2dbcTypes.RESULT_FLUX_MAPPER);
        var rowMapper = mappings.getMapping(R2dbcTypes.ROW_MAPPER);
        if (resultFluxMapper == null && rowMapper == null) {
            if (CommonUtils.isVoid(returnType)) {
                return Optional.empty();
            }
        }

        if (isMono || isFlux) {
            var publisherParam = Visitors.visitDeclaredType(returnType, dt -> dt.getTypeArguments().get(0));
            if (CommonUtils.isVoid(publisherParam)) {
                return Optional.empty();
            }
            if (publisherParam.toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
                return Optional.empty();
            }

            var mapperType = ParameterizedTypeName.get(R2dbcTypes.RESULT_FLUX_MAPPER, TypeName.get(publisherParam), TypeName.get(returnType));
            if (resultFluxMapper != null) {
                return Optional.of(new DbUtils.Mapper(resultFluxMapper.mapperClass(), mapperType, resultFluxMapper.mapperTags()));
            }
            if (rowMapper != null) {
                if (isFlux) {
                    return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.flux($L)", R2dbcTypes.RESULT_FLUX_MAPPER, c)));
                } else if (CommonUtils.isList(publisherParam)) {
                    return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.monoList($L)", R2dbcTypes.RESULT_FLUX_MAPPER, c)));
                } else if (CommonUtils.isOptional(publisherParam)) {
                    return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.monoOptional($L)", R2dbcTypes.RESULT_FLUX_MAPPER, c)));
                } else {
                    return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.mono($L)", R2dbcTypes.RESULT_FLUX_MAPPER, c)));
                }
            }
            return Optional.of(new DbUtils.Mapper(mapperType, mappings.mapperTags()));
        }

        var monoParam = TypeName.get(returnType).box();
        var mapperType = ParameterizedTypeName.get(R2dbcTypes.RESULT_FLUX_MAPPER, monoParam, ParameterizedTypeName.get(CommonClassNames.mono, monoParam));
        if (resultFluxMapper != null) {
            return Optional.of(new DbUtils.Mapper(resultFluxMapper.mapperClass(), mapperType, resultFluxMapper.mapperTags()));
        }

        if (rowMapper != null) {
            if (CommonUtils.isList(returnType)) {
                return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.monoList($L)", R2dbcTypes.RESULT_FLUX_MAPPER, c)));
            } else if (CommonUtils.isOptional(returnType)) {
                return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.monoOptional($L)", R2dbcTypes.RESULT_FLUX_MAPPER, c)));
            } else {
                return Optional.of(new DbUtils.Mapper(rowMapper.mapperClass(), mapperType, rowMapper.mapperTags(), c -> CodeBlock.of("$T.mono($L)", R2dbcTypes.RESULT_FLUX_MAPPER, c)));
            }
        }
        if (CommonUtils.isVoid(returnType)) {
            return Optional.empty();
        }
        if (returnType.toString().equals(DbUtils.UPDATE_COUNT.canonicalName())) {
            return Optional.empty();
        }

        return Optional.of(new DbUtils.Mapper(mapperType, mappings.mapperTags()));
    }

    private void enrichWithExecutor(TypeElement repositoryElement, TypeSpec.Builder builder, MethodSpec.Builder constructorBuilder) {
        builder.addField(R2dbcTypes.CONNECTION_FACTORY, "_connectionFactory", Modifier.PRIVATE, Modifier.FINAL);
        builder.addSuperinterface(R2dbcTypes.R2DBC_REPOSITORY);
        builder.addMethod(MethodSpec.methodBuilder("getR2dbcConnectionFactory")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addAnnotation(Override.class)
            .returns(R2dbcTypes.CONNECTION_FACTORY)
            .addCode("return this._connectionFactory;")
            .build());

        var executorTag = DbUtils.getTag(repositoryElement);
        if (executorTag != null) {
            constructorBuilder.addParameter(ParameterSpec.builder(R2dbcTypes.CONNECTION_FACTORY, "_connectionFactory").addAnnotation(AnnotationSpec.builder(Tag.class).addMember("value", executorTag).build()).build());
        } else {
            constructorBuilder.addParameter(R2dbcTypes.CONNECTION_FACTORY, "_connectionFactory");
        }
        constructorBuilder.addStatement("this._connectionFactory = _connectionFactory");
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/r2dbc/R2DbcRepositoryGenerator.kt
package ru.tinkoff.kora.database.symbol.processor.r2dbc

import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSFunction
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.database.symbol.processor.DbUtils
import ru.tinkoff.kora.database.symbol.processor.DbUtils.addMapper
import ru.tinkoff.kora.database.symbol.processor.DbUtils.asFlow
import ru.tinkoff.kora.database.symbol.processor.DbUtils.awaitSingle
import ru.tinkoff.kora.database.symbol.processor.DbUtils.awaitSingleOrNull
import ru.tinkoff.kora.database.symbol.processor.DbUtils.findQueryMethods
import ru.tinkoff.kora.database.symbol.processor.DbUtils.operationName
import ru.tinkoff.kora.database.symbol.processor.DbUtils.parseExecutorTag
import ru.tinkoff.kora.database.symbol.processor.DbUtils.queryMethodBuilder
import ru.tinkoff.kora.database.symbol.processor.DbUtils.resultMapperName
import ru.tinkoff.kora.database.symbol.processor.DbUtils.updateCount
import ru.tinkoff.kora.database.symbol.processor.Mapper
import ru.tinkoff.kora.database.symbol.processor.QueryWithParameters
import ru.tinkoff.kora.database.symbol.processor.RepositoryGenerator
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameter
import ru.tinkoff.kora.database.symbol.processor.model.QueryParameterParser
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.AnnotationUtils.isAnnotationPresent
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.FieldFactory
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlow
import ru.tinkoff.kora.ksp.common.FunctionUtils.isList
import ru.tinkoff.kora.ksp.common.FunctionUtils.isSuspend
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import ru.tinkoff.kora.ksp.common.parseMappingData


class R2DbcRepositoryGenerator(val resolver: Resolver) : RepositoryGenerator {
    private val repositoryInterface = resolver.getClassDeclarationByName(resolver.getKSNameFromString(R2dbcTypes.repository.canonicalName))?.asStarProjectedType()

    override fun generate(repositoryType: KSClassDeclaration, typeBuilder: TypeSpec.Builder, constructorBuilder: FunSpec.Builder): TypeSpec {
        this.enrichWithExecutor(repositoryType, typeBuilder, constructorBuilder)
        val repositoryResolvedType = repositoryType.asStarProjectedType()
        val resultMappers = FieldFactory(typeBuilder, constructorBuilder, "_result_mapper_")
        val parameterMappers = FieldFactory(typeBuilder, constructorBuilder, "_parameter_mapper_")

        var methodCounter = 1
        for (method in repositoryType.findQueryMethods()) {
            val methodType = method.asMemberOf(repositoryResolvedType)
            val parameters = QueryParameterParser.parse(R2dbcTypes.connection, R2dbcTypes.parameterColumnMapper, method, methodType)
            val queryAnnotation = method.findAnnotation(DbUtils.queryAnnotation)!!
            val queryString = queryAnnotation.findValue<String>("value")!!
            val query = QueryWithParameters.parse(queryString, parameters, method)
            val resultMapper = this.parseResultMapper(method, parameters, methodType)?.let { resultMappers.addMapper(it) }
            DbUtils.parseParameterMappers(method, parameters, query, R2dbcTypes.parameterColumnMapper) { R2dbcNativeTypes.findNativeType(it.toTypeName()) != null }
                .forEach { parameterMappers.addMapper(it) }
            val methodSpec = this.generate(typeBuilder, methodCounter, method, methodType, query, parameters, resultMapper, parameterMappers)
            typeBuilder.addFunction(methodSpec)
            methodCounter++
        }

        return typeBuilder.primaryConstructor(constructorBuilder.build()).build()
    }

    private data class QueryReplace(val index: Int, val sqlIndex: Int, val name: String)

    private fun generate(
        typeBuilder: TypeSpec.Builder,
        methodNumber: Int,
        method: KSFunctionDeclaration,
        function: KSFunction,
        query: QueryWithParameters,
        parameters: List<QueryParameter>,
        resultMapperName: String?,
        parameterMappers: FieldFactory
    ): FunSpec {
        var sql = query.rawQuery

        val replaceParams = query.parameters.asSequence()
            .flatMap { p ->
                val replaces = mutableListOf<QueryReplace>()
                for (i in p.queryIndexes.indices) {
                    val queryIndex = p.queryIndexes.get(i)
                    val sqlIndex = p.sqlIndexes.get(i)
                    replaces.add(QueryReplace(queryIndex.start, sqlIndex, p.sqlParameterName))
                }
                replaces.asSequence()
            }
            .sortedBy { it.index }
            .toList()
        var sqlIndexDiff = 0
        for (parameter in replaceParams) {
            val queryIndexAdjusted: Int = parameter.index - sqlIndexDiff
            val index: Int = parameter.sqlIndex + 1
            sql = sql.substring(0, queryIndexAdjusted) + "$" + index + sql.substring(queryIndexAdjusted + parameter.name.length + 1)
            sqlIndexDiff += (parameter.name.length - index.toString().length)
        }

        val b = method.queryMethodBuilder(resolver)

        val queryContextFieldName = "_queryContext_$methodNumber"
        typeBuilder.addProperty(
            PropertySpec.builder(queryContextFieldName, DbUtils.queryContext, KModifier.PRIVATE)
                .initializer(
                    """
                    %T(
                      %S,
                      %S,
                      %S
                    )
                    """.trimIndent(), DbUtils.queryContext, query.rawQuery, sql, method.operationName()
                )
                .build()
        )
        b.addStatement("val _query = %L", queryContextFieldName)

        val batchParam = parameters.firstOrNull { it is QueryParameter.BatchParameter }
        val returnType = function.returnType!!
        val isSuspend = method.isSuspend()
        val isFlow = method.isFlow()
        b.addCode("return ")
        b.controlFlow("%T.deferContextual { _reactorCtx ->", if (isFlow) CommonClassNames.flux else CommonClassNames.mono) {
            b.addStatement("val _ctxCurrent = %T.current(_reactorCtx)", CommonClassNames.contextReactor)
            b.addStatement("val _ctxFork = _ctxCurrent.fork()")
            b.addStatement("_ctxFork.inject()")
            b.addStatement("val _telemetry = this._r2dbcConnectionFactory.telemetry().createContext(_ctxFork, _query)")
            b.controlFlow("_r2dbcConnectionFactory.withConnection%L { _con ->", if (isFlow) "Flux" else "") {
                b.addStatement("val _stmt = _con.createStatement(_query.sql())")
                R2dbcStatementSetterGenerator.generate(b, query, parameters, batchParam, parameterMappers)
                b.addStatement("val _flux = %T.from<%T>(_stmt.execute())", CommonClassNames.flux, R2dbcTypes.result)
                if (returnType == resolver.builtIns.unitType) {
                    b.addCode("_flux.flatMap { it.rowsUpdated }.then().thenReturn(Unit)")
                } else if (returnType.toTypeName() == updateCount) {
                    if (isFlow) {
                        b.addCode("_flux.flatMap { it.rowsUpdated }.map { %T(it) }", updateCount)
                    } else {
                        b.addCode("_flux.flatMap { it.rowsUpdated }.reduce(0L) { v1, v2 -> v1 + v2 }.map { %T(it) }", updateCount)
                    }
                } else {
                    b.addCode("%N.apply(_flux)", resultMapperName)
                }
                b.controlFlow(".doOnEach { _s ->") {
                    b.controlFlow("if (_s.isOnComplete)") {
                        b.addStatement("_telemetry.close(null)")
                        b.addStatement("_ctxCurrent.inject()")
                        b.nextControlFlow("else if (_s.isOnError)")
                        b.addStatement("_telemetry.close(_s.throwable)")
                        b.addStatement("_ctxCurrent.inject()")
                    }
                }
            }
        }

        if (isSuspend) {
            if (returnType.isMarkedNullable) {
                b.addCode(".%M()\n", awaitSingleOrNull)
            } else {
                b.addCode(".%M()\n", awaitSingle)
            }
        } else if (isFlow) {
            b.addCode(".%M()\n", asFlow)
        } else {
            if (returnType.isMarkedNullable) {
                b.addCode(".block()")
            } else {
                b.addCode(".block()!!")
            }
        }
        return b.build()
    }

    private fun parseResultMapper(method: KSFunctionDeclaration, parameters: List<QueryParameter>, methodType: KSFunction): Mapper? {
        val returnType = methodType.returnType!!
        val returnTypeName = returnType.toTypeName().copy(false)
        if (returnTypeName == UNIT) {
            return null
        }
        if (returnTypeName == updateCount) {
            return null
        }

        val isGeneratedKeys = method.isAnnotationPresent(DbUtils.idAnnotation)
        for (parameter in parameters) {
            if (parameter is QueryParameter.BatchParameter) {
                if (!isGeneratedKeys) {
                    throw ProcessingErrorException("@Batch method can't return arbitrary values, it can only return: void/UpdateCount or database-generated @Id", method)
                }
            }
        }

        val mapperName = method.resultMapperName()
        val mappings = method.parseMappingData()
        val resultSetMapper = mappings.getMapping(R2dbcTypes.resultFluxMapper)
        val rowMapper = mappings.getMapping(R2dbcTypes.rowMapper)
        if (method.isFlow()) {
            val flowParam = returnType.arguments[0]
            val flowReturnTypeName = flowParam.toTypeName().copy(false)
            val mapperType = R2dbcTypes.resultFluxMapper.parameterizedBy(flowReturnTypeName, CommonClassNames.flux.parameterizedBy(flowReturnTypeName))
            if (rowMapper != null) {
                return Mapper(rowMapper, mapperType, mapperName) {
                    CodeBlock.of("%T.flux(%L)", R2dbcTypes.resultFluxMapper, it)
                }
            }
            return Mapper(mapperType, mapperName)
        }

        val mapperType = R2dbcTypes.resultFluxMapper.parameterizedBy(returnTypeName, CommonClassNames.mono.parameterizedBy(returnTypeName))
        if (resultSetMapper != null) {
            return Mapper(resultSetMapper, mapperType, mapperName)
        }

        if (rowMapper != null) {
            return if (method.isList()) {
                Mapper(rowMapper, mapperType, mapperName) {
                    CodeBlock.of("%T.monoList(%L)", R2dbcTypes.resultFluxMapper, it)
                }
            } else {
                Mapper(rowMapper, mapperType, mapperName) {
                    CodeBlock.of("%T.mono(%L)", R2dbcTypes.resultFluxMapper, it)
                }
            }
        }

        return Mapper(mapperType, mapperName)
    }

    override fun repositoryInterface() = repositoryInterface


    private fun enrichWithExecutor(repositoryElement: KSClassDeclaration, builder: TypeSpec.Builder, constructorBuilder: FunSpec.Builder) {
        builder.addProperty("_r2dbcConnectionFactory", R2dbcTypes.connectionFactory, KModifier.PRIVATE)
        builder.addSuperinterface(R2dbcTypes.repository)
        builder.addFunction(
            FunSpec.builder("getR2dbcConnectionFactory")
                .addModifiers(KModifier.OVERRIDE)
                .returns(R2dbcTypes.connectionFactory)
                .addStatement("return this._r2dbcConnectionFactory")
                .build()
        )
        val executorTag = repositoryElement.parseExecutorTag()
        if (executorTag != null) {
            constructorBuilder.addParameter(
                ParameterSpec.builder("_2dbcConnectionFactory", R2dbcTypes.connectionFactory).addAnnotation(
                    AnnotationSpec.builder(CommonClassNames.tag).addMember("value = %L", executorTag).build()
                ).build()
            )
        } else {
            constructorBuilder.addParameter("_r2dbcConnectionFactory", R2dbcTypes.connectionFactory)
        }
        constructorBuilder.addStatement("this._r2dbcConnectionFactory = _r2dbcConnectionFactory")
    }
}
